File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbstm03a.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java reimplementation of CardDemo batch program CBSTM03A — Account Statement Generator.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbstm03a.py} byte-for-byte. Python is the
 * source of truth.
 */
public class Cbstm03a implements ProgramRunner {

    private static final BigDecimal BD_0_00 = new BigDecimal("0.00");

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenarioName = inputs.getOrDefault("SCENARIO", "PROCESS_RECORDS").toUpperCase();
        Scenario scn = buildScenario(scenarioName);
        if (scn == null) {
            Map<String, String> err = new LinkedHashMap<>();
            err.put("error", "unknown scenario: '" + scenarioName + "'");
            return err;
        }

        // Build file manager with transactions
        Cbstm03b.FileManager fileMgr = new Cbstm03b.FileManager(scn.transactions);

        // OPEN, READ all, CLOSE -- group by card number
        Cbstm03b.M03BArea area = new Cbstm03b.M03BArea();
        area.oper = "O";
        fileMgr.execute(area);

        Map<String, List<TranData>> transactionsByCard = new LinkedHashMap<>();
        int totalRecordsRead = 0;

        while (true) {
            area.oper = "R";
            fileMgr.execute(area);
            if ("10".equals(area.rc)) break;
            totalRecordsRead++;

            TranData tran = parseFldt(area.fldt);
            if (tran == null) continue;

            transactionsByCard.computeIfAbsent(tran.tranCardNum, k -> new ArrayList<>()).add(tran);
        }

        area.oper = "C";
        fileMgr.execute(area);

        // Generate statements
        List<String> plainTextLines = new ArrayList<>();
        List<String> htmlLines = new ArrayList<>();
        int statementsProduced = 0;

        for (Map.Entry<String, List<TranData>> entry : transactionsByCard.entrySet()) {
            String cardNum = entry.getKey();
            List<TranData> transList = entry.getValue();
            StatementData stmtData = buildStatementData(cardNum, transList, scn);
            plainTextLines.addAll(formatPlainText(stmtData));
            htmlLines.addAll(formatHtml(stmtData));
            statementsProduced++;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("STATEMENTS_PRODUCED", String.valueOf(statementsProduced));
        out.put("TOTAL_RECORDS_READ", String.valueOf(totalRecordsRead));
        out.put("PLAIN_TEXT_LINE_COUNT", String.valueOf(plainTextLines.size()));
        out.put("HTML_LINE_COUNT", String.valueOf(htmlLines.size()));
        for (int i = 0; i < plainTextLines.size(); i++) {
            out.put("PLAIN_" + i, plainTextLines.get(i));
        }
        for (int i = 0; i < htmlLines.size(); i++) {
            out.put("HTML_" + i, htmlLines.get(i));
        }
        return out;
    }

    private TranData parseFldt(String fldt) {
        if (fldt == null || fldt.length() < 16) return null;
        try {
            TranData t = new TranData();
            t.tranId = fldt.substring(0, 16).trim();
            t.tranTypeCd = fldt.substring(16, 18).trim();
            t.tranCatCd = Integer.parseInt(fldt.substring(18, 22).trim().isEmpty() ? "0" : fldt.substring(18, 22).trim());
            t.tranSource = fldt.substring(22, 32).trim();
            t.tranDesc = fldt.substring(32, 132).trim();
            t.tranAmt = new BigDecimal(fldt.substring(132, 146).trim().isEmpty() ? "0" : fldt.substring(132, 146).trim());
            t.tranMerchantId = Integer.parseInt(fldt.substring(146, 155).trim().isEmpty() ? "0" : fldt.substring(146, 155).trim());
            t.tranMerchantName = fldt.substring(155, 205).trim();
            t.tranMerchantCity = fldt.substring(205, 255).trim();
            t.tranMerchantZip = fldt.substring(255, 265).trim();
            t.tranCardNum = fldt.substring(265, 281).trim();
            t.tranOrigTs = fldt.substring(281, 307).trim();
            t.tranProcTs = fldt.substring(307, 333).trim();
            return t;
        } catch (Exception e) {
            return null;
        }
    }

    private StatementData buildStatementData(String cardNum, List<TranData> trans, Scenario scn) {
        StatementData stmt = new StatementData();
        stmt.cardNum = cardNum;
        CardXrefRecord xref = scn.xrefs.get(cardNum);
        if (xref != null) {
            stmt.acctId = xref.xrefAcctId;
            stmt.account = scn.accounts.get(xref.xrefAcctId);
            stmt.customer = scn.customers.get(xref.xrefCustId);
        }
        stmt.transactions = trans;
        BigDecimal total = BD_0_00;
        for (TranData t : trans) {
            total = total.add(t.tranAmt);
        }
        stmt.totalAmount = total;
        return stmt;
    }

    private List<String> formatPlainText(StatementData stmt) {
        List<String> lines = new ArrayList<>();
        StringBuilder stars31 = new StringBuilder();
        for (int i = 0; i < 31; i++) stars31.append('*');
        StringBuilder stars32 = new StringBuilder();
        for (int i = 0; i < 32; i++) stars32.append('*');

        lines.add(stars31.toString() + "START OF STATEMENT" + stars31.toString());

        CustomerRecord cust = stmt.customer;
        if (cust != null) {
            String name = cust.custFirstName.trim() + " " + cust.custLastName.trim();
            lines.add(String.format("%-75s     ", name));
            lines.add(String.format("%-50s              ", cust.custAddrLine1));
            lines.add(String.format("%-50s              ", cust.custAddrLine2));
            String addr3 = cust.custAddrStateCd + " " + cust.custAddrZip;
            lines.add(String.format("%-80s", addr3));
        } else {
            lines.add("Card: " + stmt.cardNum);
        }

        StringBuilder dashes80 = new StringBuilder();
        for (int i = 0; i < 80; i++) dashes80.append('-');

        lines.add(dashes80.toString());
        // 33 spaces + "Basic Details" + 34 spaces
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 33; i++) sb.append(' ');
        sb.append("Basic Details");
        for (int i = 0; i < 34; i++) sb.append(' ');
        lines.add(sb.toString());

        // 40 spaces padding
        StringBuilder pad40 = new StringBuilder();
        for (int i = 0; i < 40; i++) pad40.append(' ');

        lines.add(String.format("Account ID         :%-20s%s", String.valueOf(stmt.acctId), pad40.toString()));

        AccountRecord acct = stmt.account;
        if (acct != null) {
            lines.add(String.format("Current Balance    :%9.2f       %s", acct.acctCurrBal.doubleValue(), pad40.toString()));
            int fico = cust != null ? cust.custFicoCreditScore : 0;
            lines.add(String.format("FICO Score         :%-20s%s", String.valueOf(fico), pad40.toString()));
        }

        lines.add(dashes80.toString());
        // 30 spaces + "TRANSACTION SUMMARY " + 30 spaces
        StringBuilder sb2 = new StringBuilder();
        for (int i = 0; i < 30; i++) sb2.append(' ');
        sb2.append("TRANSACTION SUMMARY ");
        for (int i = 0; i < 30; i++) sb2.append(' ');
        lines.add(sb2.toString());
        lines.add(dashes80.toString());
        lines.add(String.format("%-16s%-51s%13s", "Tran ID", "Tran Details", "  Tran Amount"));

        for (TranData tran : stmt.transactions) {
            String detail = tran.tranDesc.length() > 49
                    ? tran.tranDesc.substring(0, 49) : tran.tranDesc;
            lines.add(String.format("%-16s %-49s$%9.2f-", tran.tranId, detail, tran.tranAmt.doubleValue()));
        }

        lines.add(String.format("%-10s%-56s$%9.2f-", "Total EXP:", "", stmt.totalAmount.doubleValue()));
        lines.add(stars32.toString() + "END OF STATEMENT" + stars32.toString());
        return lines;
    }

    private List<String> formatHtml(StatementData stmt) {
        List<String> lines = new ArrayList<>();
        lines.add("<!DOCTYPE html>");
        lines.add("<html><head><title>CardDemo Statement</title></head><body>");
        lines.add("<table border='1'>");
        lines.add("<tr><th>Card Number</th><td>" + stmt.cardNum + "</td></tr>");
        lines.add("<tr><th>Account ID</th><td>" + stmt.acctId + "</td></tr>");

        if (stmt.account != null) {
            lines.add(String.format("<tr><th>Balance</th><td>$%.2f</td></tr>", stmt.account.acctCurrBal.doubleValue()));
        }

        lines.add("</table>");
        lines.add("<table border='1'><tr><th>Tran ID</th><th>Description</th><th>Amount</th></tr>");
        for (TranData tran : stmt.transactions) {
            String desc = tran.tranDesc.length() > 60
                    ? tran.tranDesc.substring(0, 60) : tran.tranDesc;
            lines.add(String.format("<tr><td>%s</td><td>%s</td><td>$%.2f</td></tr>",
                    tran.tranId, desc, tran.tranAmt.doubleValue()));
        }
        lines.add(String.format("<tr><td colspan='2'><b>TOTAL</b></td><td>$%.2f</td></tr>",
                stmt.totalAmount.doubleValue()));
        lines.add("</table></body></html>");
        return lines;
    }

    // ── Scenarios ─────────────────────────────────────────────────────────

    private Scenario buildScenario(String name) {
        switch (name) {
            case "PROCESS_RECORDS": return scenarioProcessRecords();
            case "EMPTY_INPUT":    return scenarioEmptyInput();
            default: return null;
        }
    }

    private Scenario scenarioProcessRecords() {
        Scenario s = new Scenario();

        s.transactions.add(new Cbstm03b.TranRecord(
                "TRN0000000000001", "01", 1,
                "", "Widget purchase",
                new BigDecimal("150.25"), 0,
                "", "", "",
                "4111000000001111",
                "2026-04-01-12.00.00.000000",
                "2026-04-01-12.00.00.000000"
        ));
        s.transactions.add(new Cbstm03b.TranRecord(
                "TRN0000000000002", "01", 1,
                "", "Gadget purchase",
                new BigDecimal("75.50"), 0,
                "", "", "",
                "4111000000001111",
                "2026-04-05-14.30.00.000000",
                "2026-04-05-14.30.00.000000"
        ));

        s.xrefs.put("4111000000001111",
                new CardXrefRecord("4111000000001111", 1, 100000001));
        s.accounts.put(100000001, new AccountRecord(
                100000001, "Y", new BigDecimal("5000.00"), new BigDecimal("10000.00")));
        s.customers.put(1, new CustomerRecord(
                1, "JOHN", "DOE", "123 MAIN ST", "APT 4B",
                "NY", "10001", 750));
        return s;
    }

    private Scenario scenarioEmptyInput() {
        return new Scenario();
    }

    // ── Inner data types ─────────────────────────────────────────────────

    private static class Scenario {
        final List<Cbstm03b.TranRecord> transactions = new ArrayList<>();
        final Map<String, CardXrefRecord> xrefs = new HashMap<>();
        final Map<Integer, AccountRecord> accounts = new HashMap<>();
        final Map<Integer, CustomerRecord> customers = new HashMap<>();
    }

    static class TranData {
        String tranId = "";
        String tranTypeCd = "";
        int tranCatCd = 0;
        String tranSource = "";
        String tranDesc = "";
        BigDecimal tranAmt = BD_0_00;
        int tranMerchantId = 0;
        String tranMerchantName = "";
        String tranMerchantCity = "";
        String tranMerchantZip = "";
        String tranCardNum = "";
        String tranOrigTs = "";
        String tranProcTs = "";
    }

    static class StatementData {
        String cardNum = "";
        int acctId = 0;
        CustomerRecord customer = null;
        AccountRecord account = null;
        List<TranData> transactions = new ArrayList<>();
        BigDecimal totalAmount = BD_0_00;
    }

    static class CardXrefRecord {
        final String xrefCardNum;
        final int xrefCustId;
        final int xrefAcctId;

        CardXrefRecord(String cardNum, int custId, int acctId) {
            this.xrefCardNum = cardNum;
            this.xrefCustId = custId;
            this.xrefAcctId = acctId;
        }
    }

    static class AccountRecord {
        final int acctId;
        final String acctActiveStatus;
        final BigDecimal acctCurrBal;
        final BigDecimal acctCreditLimit;

        AccountRecord(int acctId, String activeStatus,
                      BigDecimal currBal, BigDecimal creditLimit) {
            this.acctId = acctId;
            this.acctActiveStatus = activeStatus;
            this.acctCurrBal = currBal;
            this.acctCreditLimit = creditLimit;
        }
    }

    static class CustomerRecord {
        final int custId;
        final String custFirstName;
        final String custLastName;
        final String custAddrLine1;
        final String custAddrLine2;
        final String custAddrStateCd;
        final String custAddrZip;
        final int custFicoCreditScore;

        CustomerRecord(int custId, String firstName, String lastName,
                        String addrLine1, String addrLine2,
                        String addrStateCd, String addrZip,
                        int ficoCreditScore) {
            this.custId = custId;
            this.custFirstName = firstName;
            this.custLastName = lastName;
            this.custAddrLine1 = addrLine1;
            this.custAddrLine2 = addrLine2;
            this.custAddrStateCd = addrStateCd;
            this.custAddrZip = addrZip;
            this.custFicoCreditScore = ficoCreditScore;
        }
    }
}

File: pipeline/reimpl/python/cbstm03a.py
"""
Reimplementation of CBSTM03A — CardDemo Account Statement Generator.

Function: Read transactions (via CBSTM03B subroutine), group by card/account,
and generate statements in both plain-text and HTML formats.

Notable COBOL features exercised (now expressed in Python):
  1. Control block addressing → objects/references
  2. ALTER and GO TO → conditional dispatch
  3. COMP and COMP-3 → int / Decimal
  4. 2-dimensional array → list[list]
  5. Call to subroutine → Cbstm03bFileManager.execute()
"""

from __future__ import annotations
from dataclasses import dataclass, field
from decimal import Decimal
from typing import Callable

from .carddemo_data import TranRecord, AccountRecord, CustomerRecord, CardXrefRecord
from .cbstm03b import Cbstm03bFileManager, M03BArea, OPER_OPEN, OPER_READ, OPER_CLOSE


# ── Statement line structures (from COBOL STATEMENT-LINES) ────────────────────

@dataclass
class StatementData:
    card_num: str = ""
    acct_id: int = 0
    customer: CustomerRecord | None = None
    account: AccountRecord | None = None
    transactions: list[TranRecord] = field(default_factory=list)
    total_amount: Decimal = Decimal("0.00")


@dataclass
class StatementResult:
    plain_text_lines: list[str] = field(default_factory=list)
    html_lines: list[str] = field(default_factory=list)
    statements_produced: int = 0
    total_records_read: int = 0


# ── Repository interfaces ──────────────────────────────────────────────────────

class XrefRepository:
    def __init__(self, xrefs: dict[str, CardXrefRecord]):
        self._by_card = xrefs

    def find_by_card(self, card_num: str) -> CardXrefRecord | None:
        return self._by_card.get(card_num)


class AccountRepository:
    def __init__(self, accounts: dict[int, AccountRecord]):
        self._accounts = accounts

    def find(self, acct_id: int) -> AccountRecord | None:
        return self._accounts.get(acct_id)


class CustomerRepository:
    def __init__(self, customers: dict[int, CustomerRecord]):
        self._customers = customers

    def find(self, cust_id: int) -> CustomerRecord | None:
        return self._customers.get(cust_id)


# ── Core processing ───────────────────────────────────────────────────────────

def generate_statements(
    file_mgr: Cbstm03bFileManager,
    xref_repo: XrefRepository,
    account_repo: AccountRepository,
    customer_repo: CustomerRepository,
    logger: Callable[[str], None] = print,
) -> StatementResult:
    """Generate account statements — mirrors CBSTM03A PROCEDURE DIVISION."""
    result = StatementResult()

    # Group transactions by card number
    area = M03BArea(dd="TRNXFILE")
    area.oper = OPER_OPEN
    file_mgr.execute(area)

    transactions_by_card: dict[str, list[TranRecord]] = {}
    save_card = ""

    while True:
        area.oper = OPER_READ
        file_mgr.execute(area)
        if area.rc == "10":  # EOF
            break
        result.total_records_read += 1

        # Deserialize from fldt
        tran = _parse_fldt(area.fldt)
        if not tran:
            continue

        if tran.tran_card_num not in transactions_by_card:
            transactions_by_card[tran.tran_card_num] = []
        transactions_by_card[tran.tran_card_num].append(tran)

    area.oper = OPER_CLOSE
    file_mgr.execute(area)

    # Generate a statement per card number
    for card_num, trans_list in transactions_by_card.items():
        stmt_data = _build_statement_data(
            card_num, trans_list, xref_repo, account_repo, customer_repo
        )
        plain_lines = _format_plain_text(stmt_data)
        html_lines = _format_html(stmt_data)
        result.plain_text_lines.extend(plain_lines)
        result.html_lines.extend(html_lines)
        result.statements_produced += 1

    return result


def _parse_fldt(fldt: str) -> TranRecord | None:
    """Parse serialized FLDT back to TranRecord."""
    if not fldt or len(fldt) < 16:
        return None
    try:
        from decimal import Decimal
        return TranRecord(
            tran_id=fldt[0:16].strip(),
            tran_type_cd=fldt[16:18].strip(),
            tran_cat_cd=int(fldt[18:22].strip() or 0),
            tran_source=fldt[22:32].strip(),
            tran_desc=fldt[32:132].strip(),
            tran_amt=Decimal(fldt[132:146].strip() or "0"),
            tran_merchant_id=int(fldt[146:155].strip() or 0),
            tran_merchant_name=fldt[155:205].strip(),
            tran_merchant_city=fldt[205:255].strip(),
            tran_merchant_zip=fldt[255:265].strip(),
            tran_card_num=fldt[265:281].strip(),
            tran_orig_ts=fldt[281:307].strip(),
            tran_proc_ts=fldt[307:333].strip(),
        )
    except Exception:
        return None


def _build_statement_data(
    card_num: str,
    trans: list[TranRecord],
    xref_repo: XrefRepository,
    account_repo: AccountRepository,
    customer_repo: CustomerRepository,
) -> StatementData:
    stmt = StatementData(card_num=card_num)
    xref = xref_repo.find_by_card(card_num)
    if xref:
        stmt.acct_id = xref.xref_acct_id
        stmt.account = account_repo.find(xref.xref_acct_id)
        stmt.customer = customer_repo.find(xref.xref_cust_id)
    stmt.transactions = trans
    stmt.total_amount = sum((t.tran_amt for t in trans), Decimal("0.00"))
    return stmt


def _format_plain_text(stmt: StatementData) -> list[str]:
    """Generate plain text statement lines (80 chars)."""
    lines: list[str] = []
    lines.append("*" * 31 + "START OF STATEMENT" + "*" * 31)

    cust = stmt.customer
    if cust:
        name = f"{cust.cust_first_name.strip()} {cust.cust_last_name.strip()}"
        lines.append(f"{name:<75}     ")
        lines.append(f"{cust.cust_addr_line_1:<50}              ")
        lines.append(f"{cust.cust_addr_line_2:<50}              ")
        addr3 = f"{cust.cust_addr_state_cd} {cust.cust_addr_zip}"
        lines.append(f"{addr3:<80}")
    else:
        lines.append(f"Card: {stmt.card_num}")

    lines.append("-" * 80)
    lines.append(" " * 33 + "Basic Details" + " " * 34)
    lines.append(f"Account ID         :{stmt.acct_id!s:<20}{' ' * 40}")

    acct = stmt.account
    if acct:
        lines.append(f"Current Balance    :{float(acct.acct_curr_bal):9.2f}       {' ' * 40}")
        fico = cust.cust_fico_credit_score if cust else 0
        lines.append(f"FICO Score         :{fico!s:<20}{' ' * 40}")

    lines.append("-" * 80)
    lines.append(" " * 30 + "TRANSACTION SUMMARY " + " " * 30)
    lines.append("-" * 80)
    lines.append(f"{'Tran ID':<16}{'Tran Details':<51}{'  Tran Amount':>13}")

    for tran in stmt.transactions:
        detail = f"{tran.tran_desc[:49]}"
        lines.append(
            f"{tran.tran_id:<16} {detail:<49}${float(tran.tran_amt):>9.2f}-"
        )

    lines.append(f"{'Total EXP:':<10}{' ' * 56}${float(stmt.total_amount):>9.2f}-")
    lines.append("*" * 32 + "END OF STATEMENT" + "*" * 32)
    return lines


def _format_html(stmt: StatementData) -> list[str]:
    """Generate HTML statement."""
    lines: list[str] = []
    lines.append("<!DOCTYPE html>")
    lines.append("<html><head><title>CardDemo Statement</title></head><body>")
    lines.append("<table border='1'>")
    lines.append(f"<tr><th>Card Number</th><td>{stmt.card_num}</td></tr>")
    lines.append(f"<tr><th>Account ID</th><td>{stmt.acct_id}</td></tr>")

    if stmt.account:
        lines.append(f"<tr><th>Balance</th><td>${float(stmt.account.acct_curr_bal):.2f}</td></tr>")

    lines.append("</table>")
    lines.append("<table border='1'><tr><th>Tran ID</th><th>Description</th><th>Amount</th></tr>")
    for tran in stmt.transactions:
        lines.append(
            f"<tr><td>{tran.tran_id}</td>"
            f"<td>{tran.tran_desc[:60]}</td>"
            f"<td>${float(tran.tran_amt):.2f}</td></tr>"
        )
    lines.append(f"<tr><td colspan='2'><b>TOTAL</b></td><td>${float(stmt.total_amount):.2f}</td></tr>")
    lines.append("</table></body></html>")
    return lines


# ── run_vector adapter ───────────────────────────────────────────────────────

def _scenario_process_records():
    from .carddemo_data import TranRecord as TR, CardXrefRecord as CXR
    trans = [
        TR(tran_id="TRN0000000000001", tran_type_cd="01", tran_cat_cd=1,
           tran_desc="Widget purchase", tran_amt=Decimal("150.25"),
           tran_card_num="4111000000001111",
           tran_orig_ts="2026-04-01-12.00.00.000000",
           tran_proc_ts="2026-04-01-12.00.00.000000"),
        TR(tran_id="TRN0000000000002", tran_type_cd="01", tran_cat_cd=1,
           tran_desc="Gadget purchase", tran_amt=Decimal("75.50"),
           tran_card_num="4111000000001111",
           tran_orig_ts="2026-04-05-14.30.00.000000",
           tran_proc_ts="2026-04-05-14.30.00.000000"),
    ]
    xrefs = {
        "4111000000001111": CXR(xref_card_num="4111000000001111",
                                xref_cust_id=1, xref_acct_id=100000001),
    }
    accounts = {
        100000001: AccountRecord(
            acct_id=100000001, acct_active_status="Y",
            acct_curr_bal=Decimal("5000.00"), acct_credit_limit=Decimal("10000.00"),
        ),
    }
    customers = {
        1: CustomerRecord(
            cust_id=1, cust_first_name="JOHN", cust_last_name="DOE",
            cust_addr_line_1="123 MAIN ST", cust_addr_line_2="APT 4B",
            cust_addr_state_cd="NY", cust_addr_zip="10001",
            cust_fico_credit_score=750,
        ),
    }
    return trans, xrefs, accounts, customers


def _scenario_empty_input():
    return [], {}, {}, {}


_CBSTM03A_SCENARIOS = {
    "PROCESS_RECORDS": _scenario_process_records,
    "EMPTY_INPUT": _scenario_empty_input,
}


def run_vector(inputs: dict) -> dict:
    """Adapter for the differential harness runner contract."""
    scenario_name = str(inputs.get("SCENARIO", "PROCESS_RECORDS")).upper()
    if scenario_name not in _CBSTM03A_SCENARIOS:
        return {"error": f"unknown scenario: {scenario_name!r}"}

    trans, xrefs, accounts, customers = _CBSTM03A_SCENARIOS[scenario_name]()
    file_mgr = Cbstm03bFileManager([
        TranRecord(
            tran_id=t.tran_id, tran_type_cd=t.tran_type_cd,
            tran_cat_cd=t.tran_cat_cd, tran_source=t.tran_source,
            tran_desc=t.tran_desc, tran_amt=t.tran_amt,
            tran_merchant_id=t.tran_merchant_id,
            tran_merchant_name=t.tran_merchant_name,
            tran_merchant_city=t.tran_merchant_city,
            tran_merchant_zip=t.tran_merchant_zip,
            tran_card_num=t.tran_card_num,
            tran_orig_ts=t.tran_orig_ts,
            tran_proc_ts=t.tran_proc_ts,
        ) for t in trans
    ])
    xref_repo = XrefRepository(xrefs)
    account_repo = AccountRepository(accounts)
    customer_repo = CustomerRepository(customers)

    result = generate_statements(file_mgr, xref_repo, account_repo, customer_repo,
                                  logger=lambda _: None)

    out: dict[str, str] = {
        "STATEMENTS_PRODUCED": str(result.statements_produced),
        "TOTAL_RECORDS_READ": str(result.total_records_read),
        "PLAIN_TEXT_LINE_COUNT": str(len(result.plain_text_lines)),
        "HTML_LINE_COUNT": str(len(result.html_lines)),
    }
    for i, line in enumerate(result.plain_text_lines):
        out[f"PLAIN_{i}"] = line
    for i, line in enumerate(result.html_lines):
        out[f"HTML_{i}"] = line
    return out
